akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala

```diff
@@ -992,6 +1034,15 @@ private[cluster] class ClusterCoreDaemon(publisher: ActorRef) extends Actor with
       }
       val newGossip = localGossip copy (members = newMembers, overview = newOverview, version = newVersion)
 
+      if (!exitingInProgress && newGossip.member(selfUniqueAddress).status == Exiting) {
+        // Leader is moving itself from Leaving to Exiting.
+        // ExitingCompleted will be received via CoordinatedShutdown to continue
+        // the leaving process. Meanwhile the gossip state is not marked as seen.
+        exitingInProgress = true
+        logInfo("Exiting (leader), starting coordinated shutdown")
+        coordShutdown.run()
```

Comment:
Hm, basically if I translate the problem, then the issue is how to collect events from "first phases", i.e. those that have no dependencies. If we have a good way for that, then this should just fire off an event that "cluster-exiting" phase is done and handle the rest declaratively. For example there can be a dedicated system actor that collects events for no dependency phases and instead of calling run() here, notify the actor that cluster is exiting, which might trigger the shutdown sequence, but might wait for other events first.

For example I can imagine a non-clustered Akka system running HTTP, that receives a graceful shutdown from the user via some external channel (pushing a button on a management page) that triggers the phase "user-shutdown" which transitively triggers the shutdown sequence. Maybe I am overthinking it though.